File: unity_code/Actions/GameActionExecutor.cs
using UnityEngine;
using System.Collections.Generic;
using System;

namespace Assets.Scripts.Actions
{
    public class GameActionExecutor : MonoBehaviour
    {
        private readonly List<IGameAction> actions = new List<IGameAction>();

        private bool isRunning = false;
        public event Action AllActionsEnd = () => {};

        public bool IsRunning
        {
            get { return isRunning; }
        }

        public void Run(IGameAction action)
        {
            actions.Add(action);
            isRunning = true;
        }

        public void Stop()
        {
            actions.Clear();
            isRunning = false;
        }

        private void Update()
        {
            if(!isRunning)
                return;

            RunActions();
        }

        private void RunActions()
        {
            var actionsToEnd = new List<IGameAction>();

            foreach(var action in actions)
            {
                var result = action.Step(Time.deltaTime);
                if(result == ActionResult.Finished)
                    actionsToEnd.Add(action);
            }

            foreach(var action in actionsToEnd)
                actions.Remove(action);

            if(actions.Count == 0)
            {
                isRunning = false;
                AllActionsEnd();
            }
        }
    }
}

File: src/main/ts/actions/GameActionExecutor.ts
import { IGameAction } from './IGameAction';
import { ActionResult } from './ActionResult';
import { IAnimatable } from './../IAnimatable';

export class GameActionExecutor implements IAnimatable {

    private actions:Array<IGameAction> = [];

    private _isRunning = false;
    public allActionsEnd:Phaser.Signal = new Phaser.Signal();

    get isRunning():boolean {
        return this._isRunning;
    }

    public run(action:IGameAction) {
        this.actions.push(action);
        this._isRunning = true;
    }

    public stop() {
        this.actions.splice(0, this.actions.length);
        this._isRunning = false;
    }

    public animate(elapsedTime:number):void {
        if(!this._isRunning) {
            return;
        }

        this.runActions(elapsedTime);
    }

    private runActions(elapsedTime:number) {
        const actionsToEnd = [];

        for (let action of this.actions) {
            const result = action.step(elapsedTime);
            if(result == ActionResult.Finished) {
                actionsToEnd.push(action);
            }
        }

        for (let action of actionsToEnd) {
            this.actions.splice(this.actions.indexOf(action), 1);
        }

        if(this.actions.length == 0) {
            this._isRunning = false;
            this.allActionsEnd.dispatch();
        }
    }
}


